### PR TITLE
Use dummy thread for a larger stack

### PR DESCRIPTION
**Pull request by dsharletg**
On Hexagon, we can't control the size of the stack of the thread which receives FastRPC calls, and that stack is tiny. To work around this, this PR adds a dummy thread with a stack we control, and forwards the calls to run to this thread.

### CODE REVIEW

File: src/runtime/hexagon_remote/halide_remote.cpp [hunk 1]
```diff
@@ -16,8 +16,96 @@ extern "C" {
...
+    qurt_mutex_t work_mutex;
+    qurt_mutex_t wakeup_master_mutex;
+
+    // Shared state
+    pipeline_argv_t function;
+    void **args;
+    int result;
+
+    void thread_main() {
+        while (true) {
```
Comment (zvookin) [inline src/runtime/hexagon_remote/halide_remote.cpp]: Per conversation, this code does not pattern match to how condition variables should be used. LEt me pull this down and do a quick rewrite as a suggestion.

File: src/runtime/hexagon_remote/halide_remote.cpp [hunk 2]
```diff
@@ -16,8 +16,96 @@ extern "C" {
...
 
+const int stack_alignment = 128;
+const int stack_size = 1024 * 1024;
+
+typedef int (*pipeline_argv_t)(void **);
+
+// We can't control the stack size on the thread which receives our
+// FastRPC calls. To work around this, we make our own thread, and
+// forward the calls to that thread.
+class PipelineContext {
```
Comment (zvookin) [inline src/runtime/hexagon_remote/halide_remote.cpp]: Here's a sketch:

`class PipelineContext {
    void *stack;
    qurt_thread_t thread;
    qurt_cond_t work_state_change;
    qurt_cond_t work_completed;
    qurt_mutex_t mutex;

    // Shared state
    bool still_running;
    pipeline_argv_t function;
    void **args;
    int result;

    void thread_main() {
        qurt_mutex_lock(&mutex);
        while (still_running) { // Favor quiting over executing pending item.
	    while (still_running && function == nullptr) {
	        qurt_cond_wait(&work_state_change, &mutex);
            }
	    if (funtion != nullptr) {
	        result = function(args);
		function = nullptr;
                qurt_cond_signal(&work_completed);
            }
        }
        qurt_mutex_unlock(&mutex);
    }

    static void redirect_main(void *data) {
        static_cast<PipelineContext *>(data)->thread_main();
    }

 +public:
    PipelineContext(int stack_alignment, int stack_size)
        : stack(NULL), still_running(true), function(NULL), args(NULL) {
        qurt_mutex_init(&work_mutex);
        qurt_mutex_init(&wakeup_master_mutex);
        qurt_cond_init(&work_state_change);
        qurt_cond_init(&work_completed);

        // Allocate the stack for this thread.
        stack = memalign(stack_alignment, stack_size);

        qurt_thread_attr_t thread_attr;
        qurt_thread_attr_init(&thread_attr);
        qurt_thread_attr_set_stack_addr(&thread_attr, stack);
        qurt_thread_attr_set_stack_size(&thread_attr, stack_size);
        qurt_thread_attr_set_priority(&thread_attr, 100);
        qurt_thread_create(&thread, &thread_attr, redirect_main, this);
    }

    ~PipelineContext() {
        still_running = false;
	qurt_cond_signal(&work_state_change);

        // Status is probably garbage. If we want it to be something, we'll need a thread exit in thread_main.
        int status;
        qurt_thread_join(thread, &status);

        free(stack);
    }

    int run(pipeline_argv_t function, void **args) {
        int local_result;

        // get a lock and set up work for the worker.
        qurt_mutex_lock(&work_mutex);
        this->function = function;
        this->args = args;

        // send a signal to the worker.
	qurt_cond_signal(&work_state_change);
	while (function != nullptr) {
	    qurt_cond_wait(&work_completed, &mutex);
        }
	local_result = result;
        qurt_mutex_unlock(&mutex);

        return local_result;
    }
};`

File: src/runtime/hexagon_remote/pipeline_context.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        qurt_cond_init(&wakeup_caller);
+
+        // Allocate the stack for this thread.
+        stack = memalign(stack_alignment, stack_size);
+
+        qurt_thread_attr_t thread_attr;
+        qurt_thread_attr_init(&thread_attr);
+        qurt_thread_attr_set_stack_addr(&thread_attr, stack);
+        qurt_thread_attr_set_stack_size(&thread_attr, stack_size);
+        qurt_thread_attr_set_priority(&thread_attr, 100);
```
Comment (abadams) [inline src/runtime/hexagon_remote/pipeline_context.h]: What does 100 mean here? Is that a lot?

Comment (dsharletg) [inline src/runtime/hexagon_remote/pipeline_context.h]: @pranavb-ca ?

File: src/runtime/hexagon_remote/pipeline_context.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    // Shared state
+    pipeline_argv_t function;
+    void **args;
+    int result;
+
+    void thread_main() {
+        while (true) {
+            // Lock and wait for work.
+            qurt_mutex_lock(&work_mutex);
+            qurt_cond_wait(&wakeup_thread, &work_mutex);
```
Comment (zvookin) [inline src/runtime/hexagon_remote/pipeline_context.h]: Per in-person discussion, I'm fairly sure there is a sleep/wakeup race here. (If the run method calls qurt_cond_signal before this executor thread gets to this wait call, the code will sleep indefinitely. Condition variables do not keep a count of signal calls. Rather if there is a thread waiting when a signal occurs, it is guaranteed to be woken. Correct code always tests the condition one is waiting for under the exclusion of the mutex before calling cond_wait. This is done in a loop as spurious wakeups are allowed. I usually think of it as if one just wrote a polling loop on some condition with a release/reacquire of the mutex in between. Then the condition variable is added to save CPU cycles. The condition variable logic guarantees atomicity between sleeping and releasing the mutex.)

**Comment by dsharletg** [comment]
Thanks @zvookin for pushing on getting the thread primitive usage correct!

**Pull request merged**